Page list Merge

Achoobert opened the pull request.

Achoobert committed:
add interface list, popup

Achoobert committed:
[add] delete and rename

wip addnewpopup menu, detail menu

Achoobert committed:
[add] parent page list 

Concerned that this will be invalid
p.constructor.name === "ABViewPage"
is replacing 
"instanceof"

Achoobert committed:
[add/fix] add new page

Achoobert committed:
[lint]

Achoobert committed:
[add] copy page

Achoobert committed:
Merge remote-tracking branch 'origin/master' into add-interface

Achoobert committed:
fix  adding interface page

make the LGTM bot happy by using 'new'
also use UI instead of AB so things match better

check if isValid is a function before trying to use it -- thi sspecifically broke adding a page to the root level OR to a child of root. Adding a page on the  ./../../___ level worked sometimes... for some reason?

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+       * remove the busy indicator from the form.
+       */
+      hide() {
+         if (this.$component) this.$component.hide();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+       * @param {ABApplication} application
+       */
+      applicationLoad(application) {
+         this.currentApplication = application; // remember our current Application.
```
super.applicationLoad()

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+       */
+      ready() {
+         if (this.$component) {
+            this.$component.hideProgress();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+           if (validator.fail()) {
+             // cb(validator); // tell current Tab component the errors
+             this[tabKey].emit("save.error", validator);
+             newInterface.label = state.old;
```
what's `state`?

Achoobert replied:
don't know, this is for the auto-generated page popup. Since it returns immediately and does nothing with that label (as far as I can tell) I removed it

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+            this.ready();
+
+            // an error happend during the server side creation.
+            // so remove this object from the current object list of
```
"object" list?

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,271 @@
+       * Show this component.
+       */
+      show() {
+         if (this.$component) this.$component.show();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage_blank.js:
```diff
@@ -0,0 +1,274 @@
+ */
+
+export default function (AB) {
+   var L = function (...params) {
```
don't forget the UI_Class method

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage_blank.js:
```diff
@@ -0,0 +1,274 @@
+         super({
+            component: base,
+
+            form: `${base}_blank`,
```
you can just leave these blank:  `form: "",`  and the super class will take care of the values.

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage_blank.js:
```diff
@@ -0,0 +1,274 @@
+       * Prepare our New Popups with the current Application
+       */
+      applicationLoad(application) {
+         this.currentApplication = application;
```
`super.applicationLoad(application);`

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_object_list.js:
```diff
@@ -110,6 +110,15 @@ export default function (AB) {
             this.ListComponent.select(obj.id);
             // }
          });
+
```
these aren't part of the way we do things now.  no more `_handler_refreshApp`

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_query_list.js:
```diff
@@ -33,6 +33,15 @@ export default function (AB) {
          });
          this.AddForm = UIListNewQuery(AB);
+
+         this._handler_refreshApp = (def) => {
```
these aren't part of the way we do things now.  no more `_handler_refreshApp`

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+/*
```
how many headers do we need?

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+ *
+ *
+ */
+
```
don't forget to implement the `UI_Class` parent

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+         super({
+            component: base,
+
+            form: `${base}_blank`,
```
leave these blank unless you really need to specify the value:  `form: "",`

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+         });
+         this.ids.parentList = {};
+
+         this.currentApplication = null;
```
handled internally now.

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+         });
+
+         // init() routines are always considered async so:
+         return Promise.resolve();
```
technically, since you declared the `fn()`  as `async`, we don't need to return a promise here.

Achoobert replied:
removed

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+       * @param {ABApplication} application
+       */
+      applicationLoad(application) {
+         this.currentApplication = application; // remember our current Application.
```
`super.applicationLoad(application);`

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+       * remove the busy indicator from the form.
+       */
+      hide() {
+         if (this.$component) this.$component.hide();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+       */
+      busy() {
+         if (this.$component) {
+            this.$component.showProgress();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+       */
+      ready() {
+         if (this.$component) {
+            this.$component.hideProgress();
```
shorthand

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+         if (values.parent === "-") {
+            values.parent = null;
+         } else if (values.parent) {
+            values.parent = this.currentApplication.urlResolve(values.parent);
```
what is the value of `values.parent` here?  I didn't think it was a url to resolve.  Is it the actual ABApplication?

Achoobert replied:
Added a comment to clarify, but after refactoring the copy function works _worse_ ... I think this will need a review later on. not sure I can catch all the issues without the ability to get into what is going on with the widgets

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+               });
+               console.log(err);
+            });
+         //await values.parent.pageInsert(values);
```
any reason to keep all these comments here?

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,453 @@
+       * Show this component.
+       */
+      show() {
+         if (this.$component) this.$component.show();
```
shorthand

Hiro-Nakamura reviewed (changes requested):
OK, got a few things to change here.

Achoobert commented on src/rootPages/Designer/ui_choose_list.js:
```diff
@@ -343,7 +343,7 @@ export default function (AB) {
 
          this._handler_reload = (def) => {
-            if (def.type == "application") {
+            if (def?.type == "application") {
```
Threw this in there, 
```
this.AB.notify.developer(new Error("No def passed"), {
    plugin: "ABDesigner",
    context: "_handler_reload(): /definition/allapplications",
 });
 ```

Achoobert replied:
Put in the error, it mostly happens on first load of the app.

Achoobert committed:
removed duplicate callback catchers

-- was getting clicks triggering double-actions (two copy popups) so I...
removed callbackPageEditMenu -- all callbacks are already handled specifically in the init
Removed matching call-backer in popup edit menu
removed onclick method of passing which popup menu item is clicked up to the importee
--bonus fixes of annoying issues--
added "?." because ids are not defined on applications first load
Removed unhelpful commented out code
'ui work interface list'
Fixed some comments on copy()

Achoobert committed:
fix change function name, remove new()

use trigger instead of confusing onItemClick
unnecessary new

Achoobert committed:
fix simple cleanup

use shorthand, get rid of comments

Achoobert commented on src/rootPages/Designer/ui_common_popupEditMenu.js:
```diff
@@ -128,29 +128,33 @@ export default function (AB) {
             //    }
             // }
 
+            if (options.onClick) {
```
I think I managed to scrub all use of onClick out of the interface.

Achoobert committed:
use UI_Class

scrub this.currentApplication
Don't pass unused variable to remove() or copy()
cleanup leftover use of process
since using UI_Class, scrub _handler_refreshApp

fold 'ready()' into 'listReady()' & busy()... since they're doing the exact same thing
replace console.errors

Achoobert commented on src/rootPages/Designer/ui_work_interface_list.js:
```diff
@@ -0,0 +1,650 @@
+         // verify they mean to do this:
+         webix.confirm({
+            title: L("Delete Page"),
+            text: L("Are you sure you wish to delete this page?", [
```
I don't understand what this means

Achoobert replied:
I think the context is renaming? but in that case the selectedPage had to exist and be valid in order to be renamed

Achoobert replied:
removed

Achoobert committed:
fix use UIClass

use UIClass
super class will take care of the values.
Clean comments
use super.applicationLoad
use shorthand

Achoobert committed:
Merge branch 'master' into add-interface

Achoobert commented on src/rootPages/Designer/ui_common_popupEditMenu.js:
```diff
@@ -179,22 +183,20 @@ export default function (AB) {
             if (option) {
-               // this._logic.callbacks.onClick(option.command);
-               this.trigger(option.command);
+                if (this.onClick) {
```
I think I was trying to avoid fetching which item was selected from the dom later: on the interface list level, when it receives a call from the popup form. It currently works, but seems inefficient. 
I removed this pass in order to get it to match the precedent though

Hiro-Nakamura reviewed (changes requested):
ok, in addition to the specified issues I wrote about,  your editor isn't paying attention to the `.eslint.rc` definitions.

There should be a 3 space tab length.  Yours is saving 2 spaces.

Make sure your environment is setup to run `eslint` + `prettier` and obey the values in the project directory.

Achoobert committed:
fix lint to 3 spaces

VScode extensions hate me

Achoobert commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,367 @@
+      var addPage = function (page, indent) {
+        indent = indent || "";
+        options.push({
+          id: page.urlPointer(),
```
It works when I use .id so using it.

Achoobert committed:
fix requested changes

removed unnecessary _handler and events
updated api for the super call
Put the CopyForm.on() into the init() method so we only call it once.
made a Label template: L("{0} - copy", [ oldName ])
Capitalized the C for current application
use page.id instead of page.urlPointer()
keep async and await Promise.all(allInits);
init doesn't need async

Achoobert commented on src/rootPages/Designer/ui_work_interface_list_newPage.js:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    switchTab(tabId) {
+      if (tabId == this.BlankTab?.ui?.body?.id) {
```
Other tab does not exist yet, refactoring this right now is out of scope.
Commented out problem and added pointer TODO for when new tab gets added

Achoobert committed:
Merge remote-tracking branch 'origin/master' into add-interface

Commit:
Merge branch 'master' into add-interface

Commit:
[fix] improper webix id

Commit:
[fix] call to unknown .ready()

Achoobert committed:
Merge remote-tracking branch 'origin/master' into add-interface

Achoobert committed:
remove unnecessary passing of label

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_copyPage.js:
```diff
@@ -0,0 +1,354 @@
+       */
+      applicationLoad(application) {
+         super.applicationLoad(application)
+         var options = [{ id: "-", value: L("[Root page]") }];
```
We mentioned in our Standup Meeting a couple of days ago about not putting [] around our text.  As it makes it look like we have a placeholder we haven't updated.

Hiro-Nakamura commented on src/rootPages/Designer/ui_work_interface_list_newPage_blank.js:
```diff
@@ -0,0 +1,280 @@
+      applicationLoad(application) {
+         super.applicationLoad(application);
+
+         var options = [{ id: "-", value: L("[Root page]") }];
```
that bracket thing

Hiro-Nakamura reviewed (changes requested):
fix a few of those bracket issues.

Also, we have implemented a Warnings interface update to the ABDesigner.  Go look at the current implementation of the Queries or Data collections and see how the Warnings are now implemented.  Go through and Implement those changes here too.

Achoobert committed:
fix no brackets

Achoobert committed:
Merge branch 'master' into add-interface

Achoobert committed:
add warnings support

Achoobert commented:
> Warnings interface update 

The workspace isn't implemented, so there are some missing things - so some of these are now commented out with TODO on them.

Hiro-Nakamura reviewed (approved):
Alright , lets merge this in and get to work.

Achoobert committed:
Merge remote-tracking branch 'origin/master' into add-interface
